### PR TITLE
fixes #24885 - proxy cloud-init nocloud templates

### PR DESCRIPTION
**Pull request by timogoebel**

### CODE REVIEW

File: modules/templates/templates_api.rb [hunk 1]
```diff
@@ -14,15 +14,21 @@ class Proxy::TemplatesApi < Sinatra::Base
     end
   end
 
+  get "/userdata/:kind" do |kind|
+    log_halt(500, "Failed to retrieve #{kind} userdata template for #{params.inspect}: ") do
+      Proxy::Templates::TemplateProxyRequest.new.get_template(['userdata', kind].join('/'), request.env, params)
```
Comment (ekohl) [inline modules/templates/templates_api.rb]: It looks like all of these join an array and inside the function it's split into an array. Shouldn't the get_template function accept an array instead?

**Comment by lzap** [comment]
Tests are red. Looks good overall.

**Comment by timogoebel** [comment]
@lzap: This is ready for another (final?) look.

**Comment by ares** [comment]
@lzap any chance you could take a look?

**Comment by lzap** [comment]
I am debugging this whole afternoon but cannot figure out why the heck I am getting 404, does it work for you?

```
2019-06-11T17:53:23 11c7b929 [I] Started GET /userdata/user-data 
2019-06-11T17:53:23 11c7b929 [D] Headers: {"HTTP_HOST"=>"localhost:8448", "HTTP_USER_AGENT"=>"curl/7.64.0", "HTTP_ACCEPT"=>"*/*", "HTTP_X_FORWARDED_FOR"=>"192.168.99.52", "HTTP_VERSION"=>"HTTP/1.1"}
2019-06-11T17:53:23 11c7b929 [I] Finished GET /userdata/user-data with 404 (0.49 ms)
2019-06-11T17:53:23  [D] close: 127.0.0.1:55972
```

The command is:

```
curl -D - -H 'X-FORWARDED-FOR: 192.168.99.52' http://localhost:8448/userdata/user-data

HTTP/1.1 404 Not Found 
Content-Type: application/json
X-Cascade: pass
Content-Length: 27
X-Content-Type-Options: nosniff
Server: foreman-proxy/1.23.0-develop
Date: Tue, 11 Jun 2019 15:53:23 GMT
Connection: Keep-Alive
```

It's not getting through, sinatra stack returns 404 immediately. Template feature is turned on and it reports back! I dunno.

File: modules/templates/proxy_request.rb [hunk 2]
```diff
@@ -0,0 +1,55 @@
...
+        raise "Unknown method: #{method}"
+      end
+      logger.debug "Retrieving a template from %s%s" % [uri, proxy_req.path]
+      logger.debug "HTTP headers: #{proxy_headers.inspect}"
+      res = send_request(proxy_req)
+      # You get a 201 from the 'built' URL
+      raise "Error retrieving #{path} for #{opts.inspect} from #{uri.host}: #{res.class}" unless ["200", "201"].include?(res.code)
+      res.body
+    end
+
```
Comment (lzap) [inline modules/templates/proxy_request.rb]: Nitpick: I'd probably move this below into private section.

**Review by lzap — APPROVED** [review]
The code looks good, except something is not right and I am getting 404.

File: modules/templates/proxy_request.rb [hunk 3]
```diff
@@ -0,0 +1,55 @@
...
+module Proxy::Templates
+  class ProxyRequest < ::Proxy::HttpRequest::ForemanRequest
+    include Proxy::Log
+
+    def call_template(method, path, env, params, body = '')
+      template_url = Proxy::Templates::Plugin.settings.template_url
+      proxy_ip = URI.parse(template_url).host
+      opts = params.clone.merge(:url => template_url)
+      opts.delete("path")
```
Comment (ekohl) [inline modules/templates/proxy_request.rb]: Would it make sense to have a `BLACKLIST_PARAMETERS = ['template', 'kind', 'hostgroup', 'splat', 'captures']` and then iterate over it inside the code?

File: modules/templates/proxy_request.rb [hunk 4]
```diff
@@ -0,0 +1,55 @@
...
+      opts.delete("kind")
+      opts.delete("hostgroup")
+      opts.delete("splat")
+      opts.delete("captures")
+      # in hostgroup provisioning there are spaces
+      path = path.map { |x| CGI.escape(x) }.join('/')
+      logger.debug "Template: request for #{path} using #{opts.inspect} at #{uri.host}"
+      proxy_headers = extract_request_headers(env)
+      proxy_headers["X-Forwarded-For"] = "#{env['REMOTE_ADDR']}, #{proxy_ip}"
+      proxy_headers["Content-Type"] = params["Content-Type"] if params["Content-Type"]
```
Comment (ekohl) [inline modules/templates/proxy_request.rb:40]: Perhaps there are more headers to pass along? `Accept` comes to mind. Also note that HTTP headers are case insensitive. I don't know if `params` is case insensitive.

**Review by ekohl — COMMENTED** [review]
After writing the comments I realized that it was just moving existing code.

**Comment by timogoebel** [comment]
@lzap, @ekohl: I checked out why this did not work. Apparently, somebody forgot to whitelist the API endpoint in `modules/templates/http_config.ru`. Darn. Fixed that now. I also incorporated your requested changes.

File: modules/templates/proxy_request.rb (hunk 4 above)
Comment (lzap) [reply modules/templates/proxy_request.rb:40]: I would not mind not doing this in this patch, but up to you @timogoebel

Comment (ekohl) [reply modules/templates/proxy_request.rb:40]: Yes, this was a comment from before I saw that it was just moving existing code around

**Review by lzap — APPROVED** [review]
And it works. Any futher comments @ekohl ?

```
2019-06-14T09:14:19  [D] accept: 127.0.0.1:53222
2019-06-14T09:14:19  [D] Rack::Handler::WEBrick is invoked.
2019-06-14T09:14:19 8a605078 [I] Started GET /userdata/user-data 
2019-06-14T09:14:19 8a605078 [D] Headers: {"HTTP_HOST"=>"localhost:8448", "HTTP_USER_AGENT"=>"curl/7.64.0", "HTTP_ACCEPT"=>"*/*", "HTTP_X_FORWARDED_FOR"=>"192.168.202.80", "HTTP_VERSION"=>"HTTP/1.1"}
2019-06-14T09:14:19 8a605078 [D] Template: request for userdata/user-data using {"url"=>"http://192.168.122.1:8448"} at 127.0.0.1
2019-06-14T09:14:19 8a605078 [D] Retrieving a template from http://127.0.0.1:3000//userdata/user-data?url=http%3A%2F%2F192.168.122.1%3A8448
2019-06-14T09:14:19 8a605078 [D] HTTP headers: {"USER_AGENT"=>"curl/7.64.0", "ACCEPT"=>"*/*", "X_FORWARDED_FOR"=>"192.168.202.80", "X-Forwarded-For"=>"127.0.0.1, 192.168.122.1"}
2019-06-14T09:14:19 8a605078 [I] Finished GET /userdata/user-data with 200 (358.33 ms)
2019-06-14T09:14:19  [D] close: 127.0.0.1:53222
```

**Review by ekohl — APPROVED** [review]
Untested, but the code looks good and I see @lzap did.

File: modules/templates/proxy_request.rb [hunk 5]
```diff
@@ -0,0 +1,56 @@
...
+      end
+      logger.debug "Retrieving a template from %s%s" % [uri, proxy_req.path]
+      logger.debug "HTTP headers: #{proxy_headers.inspect}"
+      res = send_request(proxy_req)
+      # You get a 201 from the 'built' URL
+      raise "Error retrieving #{path} for #{opts.inspect} from #{uri.host}: #{res.class}" unless ["200", "201"].include?(res.code)
+      res.body
+    end
+
+    def extract_request_headers(env)
```
Comment (ekohl) [inline modules/templates/proxy_request.rb]: Moving this into a private section actually breaks the tests that this method has

**Review by ekohl — CHANGES_REQUESTED** [review]
Oh, one more thing

File: modules/templates/proxy_request.rb (hunk 5 above)
Comment (timogoebel) [reply modules/templates/proxy_request.rb]: Yep, I know. Will move this back to non-private in a sec.

Comment (timogoebel) [reply modules/templates/proxy_request.rb]: Done.

**Comment by timogoebel** [comment]
@lzap, @ekohl: Fixed the tests and rebased. Ready to merge imho. ;-)

**Comment by lzap** [comment]
Re-tested and merged. Thanks so much for extra effort made @timogoebel this is now officially part of core and if this breaks it's on me.

**Pull request merged**